### DISCUSSION TITLE
Can anyone provide an example of inputting and outputting dynamic content?

### DISCUSSION THREAD

**Discussion by GoodManWEN** [PyO3/pyo3#2332] [Questions]
Hi everyone, I worked with this program about two years ago, and after I recently refocused on it, I noticed there's some considerable change from the way I used it before. In my previous experience, I needed to introduce `wrap_pyfunction` and write `py: Python` as the first argument of all python functions, 
```
use pyo3::prelude::*;
use pyo3::wrap_pyfunction;

#[pyfunction]
fn test(py: Python) -> () {}

#[pymodule]
fn testlib(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(test, m)?)?;
    Ok(())
}
```
Now that seems they've been removed, some of the changes are confusing to me (someone who hasn't written any rust code in the past two years, let's say look like you need to add `PyResult` to all return value now). I am wondering if someone can provide a standard `lib.rs` example that teaches me how to input python data structures into rust, I need to input `list` (one dimensional & two dimensional), bytes and strings, and again I need rust to be able to return them. Unfortunately, these are dynamic data that cannot be determined in length before input, and I cannot understand well the process how pyo3 transforms these dynamic data currently.

Also maturin is a new tool to me that doesn't require us to run `cargo build --release` like we did before, I would like to know if used in a production environment whether the `maturin develop` command is compiling the program under `--release`. 

Can anyone give any help? Thanks very much!

**Comment by davidhewitt** [comment]
For code examples, you might want to look at some of the crates listed in PyO3's README: https://github.com/PyO3/pyo3#examples

If you want a release build with maturin, you need to run `maturin develop --release`. Or alternatively you can set up a `pyproject.toml` like the one seen at https://github.com/PyO3/maturin#source-distribution, and then you can just do `pip install .` to get a release build.

**Reply by GoodManWEN** [reply]
Dude, the examples are a bit too complicated for me, and many of them don't seem to be based on the latest versions.

**Comment by davidhewitt** [comment]
> let's say look like you need to add `PyResult` to all return value now

Can you show me where in the documentation you got that impression? Wherever you can return `PyResult<T>`, `T` should also work just as well.
